esp/esp/program/models/class_.py

```diff
@@ -1111,7 +1115,27 @@ def firstBlockEvent(self):
     def isFull(self, ignore_changes=False, webapp=False):
         if len(self.get_meeting_times()) == 0:
             return True
-        if webapp and Tag.getBooleanTag('count_checked_in_only', program = self.parent_program, default = False):
+
+        now = datetime.datetime.now()
+        switch_time = None
+        if Tag.getProgramTag('switch_time_program_attendance', program=self.parent_program):
+            try:
+                switch_hour, switch_min = [int(x) for x in Tag.getProgramTag('switch_time_program_attendance', program=self.parent_program).split(":")]
+                switch_time = datetime.datetime(now.year, now.month, now.day, switch_hour, switch_min)
+            # TODO: Is this the best way to deal with someone specifying the tag incorrectly?
+            except ValueError:
+                pass
+        switch_lag = None
+        if Tag.getProgramTag('switch_lag_class_attendance', program = self.parent_program):
+            try:
+                switch_lag = int(Tag.getProgramTag('switch_lag_class_attendance', program = self.parent_program))
+            except ValueError:
+                pass
+        # TODO: Don't call num_students_attending() twice, but also don't call it unless needed?
+        if webapp and switch_lag and now >= (self.start_time_prefetchable() + timedelta(minutes=switch_lag)) and self.num_students_attending() >= 1:
+            num_students = self.num_students_attending()
+        # TODO: Don't count the number of students attending the program for every class...
+        elif webapp and switch_time and now >= switch_time and ESPUser.objects.filter(record__event="attended", record__program=self.parent_program).distinct().count() >= 1:
```

Comment:
Is the idea that you only want to see if any of the students that are enrolled in this class are attending the program? If that's the case, you could just put `id__in=self.students['enrolled']` in the filter call, or something like that.

Reply:
Nah so basically I only want to switch to program attendance numbers if some program attendance has been inputted (guarding against the event that a program wants to switch to program attendance at 10am but something goes wrong on day-of and they haven't inputted any attendance by then and then every class appears empty). Right now, when it checks whether **each class** is full, i check if there are any students marked as attending the **entire program** (the same call over and over again for each class).

An idea is maybe to add an attribute to the parent_program object like parent_program.any_attending() which would be False until there is at least one program attendance record and would only update if it's currently False & the records are changed? Is that possible? Do you have other ideas for how to check that at least 1 program attendance has been inputted without naively counting over and over?

Reply:
I don't _think_ it's possible to only change the attribute when the number of records changes from 0 to 1. The only way  I can think of to cache such an attribute/function would check the number of records _every_ time the records are changed (but I'm also not an expert on our caching practices). Ultimately, I imagine `Record.objects.filter(event="attended", program=self.parent_program).exists()` would be a pretty quick DB query, but one possible false positive would be if there are records from when registration was being tested.